bigframes/core/compile/api.py

```diff
@@ -0,0 +1,72 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import annotations
+
+from typing import Mapping, Tuple, TYPE_CHECKING
+
+import bigframes.core.compile.compiler as compiler
+
+if TYPE_CHECKING:
+    import bigframes.core.nodes
+    import bigframes.core.ordering
+    import bigframes.core.schema
+
+
+def compile_peek(node: bigframes.core.nodes.BigFrameNode, n_rows: int) -> str:
+    """Compile node into sql that selects N arbitrary rows, may not execute deterministically."""
+    return compiler.compile_unordered_ir(node).peek_sql(n_rows)
+
+
+def compile_unordered(
+    node: bigframes.core.nodes.BigFrameNode, *, col_id_overrides: Mapping[str, str] = {}
+) -> str:
+    """Compile node into sql where rows are unsorted, and no ordering information is preserved."""
+    return compiler.compile_unordered_ir(node).to_sql(col_id_overrides=col_id_overrides)
+
+
+def compile_ordered(
+    node: bigframes.core.nodes.BigFrameNode, *, col_id_overrides: Mapping[str, str] = {}
+) -> str:
+    """Compile node into sql where rows are sorted with ORDER BY."""
+    return compiler.compile_ordered_ir(node).to_sql(
+        col_id_overrides=col_id_overrides, sorted=True
+    )
+
+
+def compile_raw(
+    node: bigframes.core.nodes.BigFrameNode,
+) -> Tuple[str, bigframes.core.ordering.ExpressionOrdering]:
+    """Compile node into sql that exposes all columns, including hidden ordering-only columns."""
+    ir = compiler.compile_ordered_ir(node)
```

Comment:
Can `compiled_raw` return an unordered version of SQL?

Reply:
No, `compile_raw` exits to return the hidden ordering columns. This is needed primarily to cache as bq table without generating row numbers. `compile_unordered` should be used if ordering information isn't needed.